experimental/Schemes/SpecOpen.jl

```diff
@@ -0,0 +1,699 @@
+
+export StructureSheafRing, scheme, domain, OO
+
+export StructureSheafElem, domain, restrictions, patches, restrict, npatches
+
+export SpecOpenMor, maps_on_patches, restriction, identity_map, preimage, generic_fractions, pullback, maximal_extension
+
+@Markdown.doc """
+    SpecOpen{BRT, BRET, RT, RET, MST} <: Scheme{BRT, BRET}
+
+Zariski open subset ``U`` of an affine scheme ``X = Spec(R)``. 
+This stores a list of generators ``f₁,…,fᵣ`` of an ideal 
+``I`` defining the complement ``Z = X ∖ U``. 
+The scheme ``X`` is referred to as the *ambient scheme* and 
+the list ``f₁,…,fᵣ`` as the *generators* for ``U``.
+"""
+@attributes mutable struct SpecOpen{BRT, BRET, RT, RET, MST} <: Scheme{BRT, BRET}
+  X::Spec{BRT, BRET, RT, RET, MST} # the ambient scheme
+  gens::Vector{RET} # a list of functions defining the complement of the open subset
+
+  # fields used for caching
+  name::String
+  patches::Vector{Spec{BRT, BRET, RT, RET, MST}}
+  intersections::Dict{Tuple{Int}, Spec{BRT, BRET, RT, RET, MST}}
+  complement::Spec{BRT, BRET, RT, RET, MST}
+
+  function SpecOpen(X::Spec{BRT, BRET, RT, RET, MST}, f::Vector{RET}; name::String="") where {BRT, BRET, RT, RET, MST}
+    for a in f
+      parent(a) == base_ring(OO(X)) || error("element does not belong to the correct ring")
+    end
+    if length(name) > 0 
+      U = new{BRT, BRET, RT, RET, MST}(X, f)
+      set_name!(U, name)
+      return U
+    end
+    return new{BRT, BRET, RT, RET, MST}(X, f)
+  end
+end
+
+@Markdown.doc """
+    ambient(U::SpecOpen)
+
+Return the ambient scheme ``X`` of a Zariski open subset ``U ⊂ X``.
+"""
+ambient(U::SpecOpen) = U.X
+
+@Markdown.doc """
+    npatches(U::SpecOpen)
+
+Return the number of generators stored for describing the complement of ``U``.
+"""
+npatches(U::SpecOpen) = length(U.gens)
+
+@Markdown.doc """
+    gens(U::SpecOpen) = U.gens
+
+Return the generators ``[f₁,…,fᵣ]`` stored for the description 
+of the complement of ``U``.
+"""
+gens(U::SpecOpen) = U.gens
+
+@Markdown.doc """
+    getindex(U::SpecOpen, i::Int) = patches(U)[i]
+
+Return the hypersurface complement of ``fᵢ`` in the 
+ambient scheme ``X`` of ``U`` where ``f₁,…,fᵣ`` are 
+the generators stored for the description of the complement 
+of ``U``.
+"""
+getindex(U::SpecOpen, i::Int) = affine_patches(U)[i]
+
+function Base.show(io::IO, U::SpecOpen)
+  if isdefined(U, :name) 
+    print(io, name(U))
+    return
+  end
+  print(io, "complement of zero locus of $(gens(U)) in $(ambient(U))")
+end
+
+@Markdown.doc """
+    function SpecOpen(
+      X::Spec{BRT, BRET, RT, RET, MST},
+      I::MPolyLocalizedIdeal{BRT, BRET, RT, RET, MST}
+    ) where {BRT, BRET, RT, RET, MST}
+
+Return the complement of the zero locus of ``I`` in ``X``.
+"""
+function SpecOpen(
+    X::Spec{BRT, BRET, RT, RET, MST},
+    I::MPolyLocalizedIdeal{BRT, BRET, RT, RET, MST}
+  ) where {BRT, BRET, RT, RET, MST}
```

Comment:
No point in specifying these type parameters if they are not used in the body of the function. Same remarks applies in a few other places.
```suggestion
function SpecOpen(X::Spec, I::MPolyLocalizedIdeal)
```

Reply:
OK, I was not quite explicit here, sorry: of course they do serve a purpose, in that they ensure the parameters of the two type instances "fit together". But you still have to check the base_ring below, and that also does that, doesn't it?

Reply:
For me the point is that the parameters of X and I should coincide. Is that a valid reason?

Reply:
I think the point is to force the parameters of `Spec` and `MPolyLocalizedIdeal` to match. Is that valid?

Reply:
Is there a way they can mismatch if the base ring check succeeds? I think the concern is that the signatures are quite hard to parse, and restricting the type has no real advantage if there is no intend to use dispatch. Or maybe we want to take advantage of dispatch to provide other methods where the parameters don't match?

Reply:
In this case, no dispatch is being used. If this really causes complexity problems, I will refrain from the signature comparison in the functions definition. I thought that this might have been even more efficient than the manual comparison  afterwards. So: Good to know.

Reply:
Performance wise both are identical. The just in time compiler creates a compiled version for each concrete instance of `Spec` and `MPolyLocalizedIdeal`.  Even `function SpecOpen(X, I)` should have the same performance (?). As far as I know it is mostly about dispatch.
The advantage of not specifying {BRT,BRET,............} is that the code looks nicer and is easier to read..

Reply:
Performance *per call* may be the same here. But it still hurts us that for each ring type, we get different instances of these types, and the JIT gets invoked again and again and again... Dropping some or all of these type parameters might thus actually improve performance.

Reply:
Essentially, the base_ring should determine all the other rings. Ideally that would be the only type parameter and every other type is computed from that parameter by a function. 
I see mainly `ZZ`, `QQ`, finite fields and number fields.
For instance for Spec(Q[x,y]) we get 
`Spec{FlintRationalField, fmpq, FmpqMPolyRing, fmpq_mpoly, MPolyPowersOfElement{FlintRationalField, fmpq, FmpqMPolyRing, fmpq_mpoly}}` 
Indeed that looks horrible. Desirable, would be 
`Spec{FlintRationalField}` but I seem to remember that this was somehow not possible due to limitations of julia. 
For example `Hecke.HermLat` also depends on 5 type parameters, even though the base ring type should determine everything else. I suppose that @thofma had a good reason when implementing it this way.
Anyway, as far as recompiling goes: We have only quite few base_ring types and usually one works with a scheme over a fixed base ring, (e.g. QQ, QQ(a), ZZ, GF(p), QQ[x]) while staying in this category no recompilation should be triggered. 
Only base change will trigger recompilation. But again that is usually to a limited number of types e.g. `ZZ --> GF(p)` or `QQ -> QQ`

Reply:
I mostly agree with @simonbrandhorst : In a specific application, we will not have a whole lot of different types of rings and I don't see the risk of encountering problems such as those that once came from the hard-coded characteristic for finite fields  in the context of the Chinese remainder theorem. 

It is true that in theory (almost) all types should be derived from the type of the base ring. In practice, however, this is not the case. For instance, I started out by storing the polynomial ring as an instance of `MPolyRing{BaseRingElemType}` only to find out that if I was using `QQ`, then `fmpq_mpoly` would be used which caused errors. The only solution was to introduce a new type parameter by itself. 

But the code has some potential for improvements as I learned from a recent discussion on slack. One paradigm seems to be not to 'bubble up' the type parameters to the surface, but to capsule them. In this case, this could for example mean that the first four parameters of `MPolyLocalizedRing{BRT, BRET, RT, RET, MST}` are really already mentioned and fixed in the concrete instance of `MST`. But then, again, what would I write for the type of a field, say `a`, in `MPolyLocalizedRing` where I know it has to be of type `RT` (the type of the ring that is being localized)? If I was capsulating that in `MST`, I would have to leave the specific type for  `a` open and unspecified. 

Maybe, this is even an option in Julia: If the constructors are all set up in a type stable way, then in practice `a` will always automatically be of the correct type. I would like to discuss such things with someone who is more an expert in julia since I'm not (yet). 

I can try to keep in mind the concerns of @fingolfin for the future developments and I agree that all this type salad is horrible to look at, but with my state of knowledge, I simply don't know how to solve all these issues in another way.

Reply:
> For instance, I started out by storing the polynomial ring as an instance of `MPolyRing{BaseRingElemType}` only to find out that if I was using `QQ`, then `fmpq_mpoly` would be used which caused errors. The only solution was to introduce a new type parameter by itself.

I don't understand, something seems to be missing here. Where was `fmpq_mpoly` used, how did it cause errors?


> But then, again, what would I write for the type of a field, say `a`, in `MPolyLocalizedRing` where I know it has to be of type `RT` (the type of the ring that is being localized)?

Well... for example `Ring`? There is often no point in specifying the *exact* type. Sure, it can help to prevent type dispatch. But trying to specify every type exactly is akin to removing multi dispatch, a core feature of Julia. Yes, sometimes you want to avoid it for performance, but in many other cases, it is fine to embrace it.

Reply:
> I simply don't know how to solve all these issues in another way.

I'd try to help but I don't know what the issues are, precisely...

Reply:
> > For instance, I started out by storing the polynomial ring as an instance of `MPolyRing{BaseRingElemType}` only to find out that if I was using `QQ`, then `fmpq_mpoly` would be used which caused errors. The only solution was to introduce a new type parameter by itself.
> 
> I don't understand, something seems to be missing here. Where was `fmpq_mpoly` used, how did it cause errors?

I declared the field in the `struct` as `a::MPolyRing{BRET}` and then the constructor tried to assign to this field the output of `PolynomialRing(QQ, [...` which was `fmpq_mpoly`. That produced an error. 

> 
> > But then, again, what would I write for the type of a field, say `a`, in `MPolyLocalizedRing` where I know it has to be of type `RT` (the type of the ring that is being localized)?
> 
> Well... for example `Ring`? There is often no point in specifying the _exact_ type. Sure, it can help to prevent type dispatch. But trying to specify every type exactly is akin to removing multi dispatch, a core feature of Julia. Yes, sometimes you want to avoid it for performance, but in many other cases, it is fine to embrace it.

This is probably something that I still need to learn: Where exactly does the thin line for type-stability run? As far as I understand, `::Ring` does nothing in terms of type stability, because the type is not concrete. So inside a `struct` it is merely a reminder to the programmer what is expected in this place. But then I need to understand better what else needs to be done in order to have type stable code. If we found some time to go through the Schemes code together, that would be great; the sooner the better.

Reply:
Just a few minor remarks (hopefully related):

1. I don't think there is a JIT performance problem. As @simonbrandhorst said the first type determines the rest of 
them so for every ring there is only one associated new type. If we don't use a parameter for this, then I don't know why we should be using julia at all :)
2. Just to get back to the example. For some imaginary type one would like to do the following:
  ```julia
  struct FancyStruct{T}
    element::T
    poly::polynomial_type(T)
  ```
  but this is not possible with the current implementation of julia. Then the next try is
```julia
struct FancyStruct{T}
  element::T
  polynomial::Generic.Poly{T}
```
but then you notice that sometimes you don't get a generic polynomial but something like an `fmpq_poly`. So you introduce the second parameter. But this is we really bad with nesting etc, so you end up with these abominations.

3. All these fancy parameters can be dropped when one introduces type stable getter/accessor functions. This is working because julia knows how to do `polynomial(F::FancyStruct{T}) = F.polynomial::polynomial_type(T)`. So one just needs to add some helper functions for the types and the getter/functions.
4. Obviously for low-level stuff one tries to not do 2. but to have concrete types for every field. 
5. The only reason why there needs to be a type would be for dispatch. So if I want to have different behavior for `FancyStruct{fmpq, fmpq_poly}` and `FancyStruct{fmpq, Generic.Poly{fmpq}`, then I need a type paramter.
6. For some types I propbably overdid it somewhere and if I were to redo them, I would now use less parameters and also less fields and more attributes.

Reply:
P.S.: I do not mind the many parameters. I don't think they make anything slow. My only problem with them is that if they are all used in signatures, the signatures become hard to parse and the docstrings quite painful. But as I said elsewhere, there is usually no reason to have them in the signature at all.

Reply:
Thanks for the explanations and clarifications. This also covers the situation I was trying to describe before. As you said, the getter-functions with type-assertion seem to be a solution to some of our problems. But I don't see this strategy backed up anywhere from the Julia community, i.e. I don't find it in any tutorials or in the documentation of the language. If I start using this strategy, I would like to make sure that it is and continues to be supported by the language and that it is not recognized as a dirty hack that will cause troubles in some future versions of Julia's compiler.

Reply:
It is used all over in julia itself, e.g.
https://github.com/JuliaLang/julia/blob/c5158f4acae792e4cd424aecaa0751b9c63005e6/stdlib/LinearAlgebra/src/symmetric.jl#L208
We don't need other reasons to start using it, because we are already using it.

Reply:
@thofma  
>     4. Obviously for low-level stuff one tries to not do 2. but to have concrete types for every field.
I do not quite get it? I thought that your conclusion was that 2. does not work for (low level) fancy types.  

>     5. The only reason why there needs to be a type would be for dispatch. So if I want to have different behavior for `FancyStruct{fmpq, fmpq_poly}` and `FancyStruct{fmpq, Generic.Poly{fmpq}`, then I need a type paramter.
You mean in case one has type-stable getters and setters?
>     6. For some types I propbably overdid it somewhere and if I were to redo them, I would now use less parameters and also less fields and more attributes.
I think a best practice example in our manual would be great.